conda/resolve.py

```diff
@@ -357,167 +325,181 @@ def verify_specs(self, specs):
+                           all(any(filter.get(f2, True) for f2 in self.find_matches(ms))
+                               for ms in self.ms_depends(fkey)))
                     filter[fkey] = sat
-                    if not sat:
-                        bad_deps.append(fkey)
+                    nnew += sat
 
-            # Build dependency chains if we detect unsatisfiability
-            nnew = nold - len(bad_deps)
             reduced = nnew < nold
             if reduced:
                 log.debug('%s: pruned from %d -> %d' % (name, nold, nnew))
-            if nnew == 0:
-                if name in snames:
-                    snames.remove(name)
-                bad_deps = [fkey for fkey in bad_deps if self.match_any(matches, fkey)]
-                matches = [(ms,) for ms in matches]
-                chains = [a + b for a in chains for b in matches] if chains else matches
-                if bad_deps:
-                    dep2 = set()
-                    for fkey in bad_deps:
-                        for ms in self.ms_depends(fkey):
-                            if not any(filter.get(f2, True) for f2 in self.find_matches(ms)):
-                                dep2.add(ms)
-                    chains = [a + (b,) for a in chains for b in dep2]
-                unsat.update(chains)
-                return nnew != 0
-            if not reduced and not first:
-                return False
+            if any(ms.optional for ms in matches):
+                return reduced
+            elif nnew == 0:
+                # Indicates that a conflict was found; we can exit early
+                return None
 
             # Perform the same filtering steps on any dependencies shared across
             # *all* packages in the group. Even if just one of the packages does
             # not have a particular dependency, it must be ignored in this pass.
-            if first:
+            if reduced or name not in snames:
                 snames.add(name)
-                if match1 not in specs:
-                    nspecs.add(MatchSpec(name))
-            cdeps = defaultdict(list)
-            for fkey in group:
-                if filter[fkey]:
-                    for m2 in self.ms_depends(fkey):
-                        if m2.name[0] != '@' and not m2.optional:
-                            cdeps[m2.name].append(m2)
-            cdeps = {mname: set(deps) for mname, deps in iteritems(cdeps) if len(deps) >= nnew}
-            if cdeps:
-                matches = [(ms,) for ms in matches]
-                if chains:
-                    matches = [a + b for a in chains for b in matches]
-                if sum(filter_group(deps, chains) for deps in itervalues(cdeps)):
-                    reduced = True
+                cdeps = {}
+                for fkey in group:
+                    if filter.get(fkey, True):
+                        for m2 in self.ms_depends(fkey):
+                            if m2.name[0] != '@' and not m2.optional:
```

Comment:
Here `m2` is a `MatchSpec` object.  Instead of this string munging, we should add a property to `MatchSpec`.


```
@property
def is_feature(self):
    return self.name[0] == '@'
```

Or whatever the `@` startswith is supposed to mean.  It's both more expressive and extensible.